source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,267 @@
+    ASSERT(slotRange.size() >= 3);
+
+    // field 0: start slot range
+    auto startField = slotRange[0].asInteger();
+
+    // field 1: end slot range
+    auto endField = slotRange[1].asInteger();
+
+    // field 2: master address for slot range
+    // TODO(hyang): for now we're only adding the master node for each slot, when we're ready to
+    //  send requests to slave nodes, we need to add subsequent address in the response as slave
+    //  nodes
+    auto masterAddress = ProcessCluster(slotRange[2]);
+    // add to the slot
+    for (auto i = startField; i <= endField; ++i) {
+      slots_[i] = masterAddress->asString();
+      // TODO: Investigate performance of alternative implementations, as this will be O(n)
+      // where n is number of shards.
+    }
+    new_hosts.emplace_back(
+        new RedisHost(parent_.info(), "", std::move(masterAddress), parent_, true));
+  }
+
+  std::unordered_map<std::string, Upstream::HostSharedPtr> updated_hosts;
+  Upstream::HostVector hosts_added;
+  Upstream::HostVector hosts_removed;
+  if (parent_.updateDynamicHostList(new_hosts, hosts_, hosts_added, hosts_removed, updated_hosts,
+                                    all_hosts_)) {
+    ASSERT(std::all_of(hosts_.begin(), hosts_.end(), [&](const auto& host) {
+      return host->priority() == locality_lb_endpoint_.priority();
+    }));
+    parent_.updateAllHosts(hosts_added, hosts_removed, locality_lb_endpoint_.priority());
+  } else {
+    parent_.info_->stats().update_no_rebuild_.inc();
+  }
+
+  all_hosts_ = std::move(updated_hosts);
+  cluster_slots_map_.swap(slots_);
+
+  // If there is an initialize callback, fire it now. Note that if the cluster refers to
+  // multiple DNS names, this will return initialized after a single DNS resolution
+  // completes. This is not perfect but is easier to code and unclear if the extra
+  // complexity is needed so will start with this.
+  parent_.onPreInitComplete();
+  resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+}
+
+void RedisCluster::RedisDiscoverySession::onFailure() {
+  current_request_ = nullptr;
+  if (client_) {
+    client_->close();
+  }
+  parent_.info()->stats().update_failure_.inc();
+  resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+}
+
+RedisCluster::ClusterSlotsRequest RedisCluster::ClusterSlotsRequest::instance_;
+
+Upstream::ClusterImplBaseSharedPtr RedisClusterFactory::createClusterWithConfig(
+    const envoy::api::v2::Cluster& cluster,
+    const envoy::config::cluster::redis::RedisClusterConfig& proto_config,
+    Upstream::ClusterFactoryContext& context,
+    Envoy::Server::Configuration::TransportSocketFactoryContext& socket_factory_context,
+    Envoy::Stats::ScopePtr&& stats_scope) {
+  if (!cluster.has_cluster_type() ||
+      cluster.cluster_type().name() != Extensions::Clusters::ClusterTypes::get().Redis) {
+    throw EnvoyException("Redis cluster can only created with redis cluster type");
```

Comment:
Please add an EXPECT_THROW_WITH_MESSAGE test for this.

Reply:
ok